Implement the index API on top of GraphQL

hannesj opened the pull request:
As discussed e.g. in #2025, it would be nice to have an API where the users could specify what information they would like to receive from the GraphIndex. Currently our front-end stack is based on React and Flux and with both of them being part of the same ecosystem as [GraphQL](http://facebook.github.io/graphql/), we thought that it would be a suitable framework for us to evaluate.

The branch uses the [graphql-java](https://github.com/andimarek/graphql-java/) implementation from @andimarek, which is still under development. The pull request is more of a discussion opener, whether this type of endpoint would be suitable for inclusion in the mainline OTP and how it should be implemented in that case. The API currently covers only a very limited set of the regular Index API, but I hope to extend that during the next days to have similar coverage to the regular Index API.

andimarek commented:
Wow ... very cool to see in this early stage a usage of grapqhl-java. As noted by @hannesj the library is under development, but if you have any questions/feedback please let me know.

abyrd commented:
@hannesj thanks for sharing this work. This could certainly be a better approach than trying to create an exhaustive set of API endpoints for the transit index. I look forward to trying this out and will give you some feedback soon.

abyrd commented:
Here are some notes for anyone who wants to test this branch:

- clone the https://github.com/andimarek/graphql-java/ repo

- modify its build.gradle to make it Maven-deployable by adding the lines

```
apply plugin: 'maven'
group = 'graphql'
```

- build and deploy graphql-java to the local repo with `./gradlew install`

-  refresh the OTP Maven model if you're running OTP in an IDE

- create a file `test.graphql` containing a sample query

```
query Agency {
  agency(id: "TriMet"){
    name
    routes {
      shortName
    }
  }
}
```

- use CURL to test out the new endpoint:

`$ curl -H "Content-Type:text/plain" -X POST http://localhost:8080/otp/routers/default/index/graphql -d @test.graphql`

abyrd commented:
@andimarek thanks for your work on the library, this looks really helpful!

abyrd commented:
```
query StopTimes {
  trip (id: "TriMet:5110179") {
    stoptimes {
      scheduledArrival
      arrivalDelay
    }
  }
}
```

This is fabulous. I am +1 on this feature.

hannesj commented:
Add 
    stop {
      name
      code
    }

inside the stoptimes, and it feels even more amazing

andimarek commented:
Great to see such positive feedback from early adopters. I'm trying to release a first version soon. This should ease your setup pain. :-) Of course any bug-reports or problems will help to improve it.

andimarek commented:
I released a first version of graphql-java. Please have a look at https://github.com/andimarek/graphql-java#maven for details how to use it with maven. I would appreciate feedback if something doesn't work for you. Thanks!

hannesj commented:
I updated this to work with the released graphql-java version and added the repository to the pom-file.

Currently still missing are:
- [ ] Stoptimes endpoint for stops (might become multiple endpoints for different needs)
- [ ] Various enums
- [ ] Checking that everything works...
- [ ] ... and writing tests, so that they will work also in the future
- [ ] Including alerts and other stuff that is not currently in the GraphIndex

It would be nice to get feedback if there is something else missing, or that you would like to get something else included.

hannesj commented:
Hi,

I added most of the missing features to the branch now. 

However, I have become a bit unsure if the graph index API is the correct place for it or if it should live one step above in the hierarchy parallel to the index API. This might not be an issue at the moment, but I'm feeling the urge to extend the GraphQL API to include stuff like regular trip planner endpoint. Currently the planner endpoint returns quite much uninteresting information and sometimes you would like to have more information (e.g. a list of routes you can transfer to at each of the stops along the way), which would be very easy with the GraphQL API, mandating the extension.

hannesj commented:
We have modified the schema a bit to suit our needs better (can be found [here](https://github.com/HSLdevcom/OpenTripPlanner/compare)).

We also provide a web interface to test the API at http://dev.hsl.fi/graphql/console/ with our current data. The interface provides autocompletion and shows possible queries and fields with ctrl+space. Please feel free to propose any changes, especially related to how we should use Connections (as defined by relay) and which queries/fields should be parametrized in a different way.

Also, we have deployed this in our beta application at http://matka.hsl.fi which now uses react-relay to query the API in an effective way. The sources for the ui can be found at https://github.com/HSLdevcom/digitransit-ui

hannesj commented:
We have now been using this for some time in our application and are happy with how it works. Probably there will be some updates in the future, but we consider this ready to be merged.

tuukka commented:
The content type has changed at some point, so here's an updated example on how to use this API from the command line: 
`
$ curl -H "Content-Type: application/graphql" -d "{agencies {gtfsId, name}}" http://digitransit.fi/otp/routers/finland/index/graphql
`

abyrd commented:
Hi @tuukka and @hannesj, I've finally merged this. There was a conflict for a while with @johannilsson's change to add feed IDs to agencies. I resolved this with some stopgap helper functions, but we'll need to come to some more uniform solution throughout OTP. I think the answer is just using feedId:entityId for absolutely everything, including Agencies, and modeling the feed scoped ID with the same class everywhere.
